test/solver/batch_bicgstab_kernels.cpp

```diff
@@ -0,0 +1,232 @@
+                .with_tolerance_type(gko::batch::stop::ToleranceType::relative)
+                .on(exec);
+        return gko::test::generate_3pt_stencil_batch_problem<MatrixType>(
+            exec, num_batch_items, num_rows, num_rhs,
+            std::forward<MatrixArgs>(args)...);
+    }
+
+    std::function<void(const Settings, const gko::batch::BatchLinOp*,
+                       const Mtx*, const MVec*, MVec*, LogData&)>
+        solve_lambda;
+    Settings solver_settings{};
+    std::shared_ptr<solver_type::Factory> solver_factory;
+};
+
+
+TEST_F(BatchBicgstab, SolvesStencilSystem)
+{
+    const int num_batch_items = 2;
+    const int num_rows = 10;
+    const int num_rhs = 1;
+    const real_type tol = 1e-5;
+    const int max_iters = 100;
+    auto linear_system = setup_linsys_and_solver<Mtx>(num_batch_items, num_rows,
+                                                      num_rhs, tol, max_iters);
+
+    auto res = gko::test::solve_linear_system(exec, solve_lambda,
+                                              solver_settings, linear_system);
+
+    for (size_t i = 0; i < num_batch_items; i++) {
+        ASSERT_LE(res.res_norm->get_const_values()[i] /
+                      linear_system.rhs_norm->get_const_values()[i],
+                  solver_settings.residual_tol);
+    }
+    GKO_ASSERT_BATCH_MTX_NEAR(res.x, linear_system.exact_sol, tol);
+}
+
+
+TEST_F(BatchBicgstab, StencilSystemLoggerLogsResidual)
+{
+    const int num_batch_items = 2;
+    const int num_rows = 10;
+    const int num_rhs = 1;
+    const real_type tol = 1e-5;
+    const int max_iters = 100;
+    auto linear_system = setup_linsys_and_solver<Mtx>(num_batch_items, num_rows,
+                                                      num_rhs, tol, max_iters);
+
+    auto res = gko::test::solve_linear_system(exec, solve_lambda,
+                                              solver_settings, linear_system);
+
+    auto res_log_array = res.log_data->res_norms.get_const_data();
+    for (size_t i = 0; i < num_batch_items; i++) {
+        ASSERT_LE(res_log_array[i] / linear_system.rhs_norm->at(i, 0, 0),
+                  solver_settings.residual_tol);
+        ASSERT_NEAR(res_log_array[i], res.res_norm->get_const_values()[i],
+                    10 * tol);
+    }
+}
+
+
+TEST_F(BatchBicgstab, StencilSystemLoggerLogsIterations)
+{
+    const int num_batch_items = 2;
+    const int num_rows = 10;
+    const int num_rhs = 1;
+    const int ref_iters = 5;
+    auto linear_system = setup_linsys_and_solver<Mtx>(num_batch_items, num_rows,
+                                                      num_rhs, 0, ref_iters);
+
+    auto res = gko::test::solve_linear_system(exec, solve_lambda,
+                                              solver_settings, linear_system);
+
+    auto iter_array = res.log_data->iter_counts.get_const_data();
+    for (size_t i = 0; i < num_batch_items; i++) {
+        ASSERT_EQ(iter_array[i], ref_iters);
+    }
```

Comment:
does it make the linear system unsolved? otherwise, it might be less than ref_iters

Reply:
Yes, the tolerance of 0 is not acheivable and it should always hit the ref iters

Reply:
using nan is maybe more general, which also fit if we decide to use `<=` not `<`

Reply:
Will that work on device as well ?

Reply:
Yes, I think so. It should work if the compiler does not use fast math.

Reply:
In this case, it is still not possible be acheive a tolerance of 0, so i think nan is not necessary.